### PR TITLE
Feature/issue 1271 analogue finite diff to autodiff

### PR DESCRIPTION
**Pull request by rtrangucci**
#### Summary:

Addresses #1271, branched off of feature/issue-1272-gradient-of-hessian-autodiff_hpp 

#### Intended Effect:

Finite difference functionals to simplify testing. Implemented:

- Gradient
- Hessian: pure finite difference, and 1st order finite difference of autodiff gradient
- Gradient of the Hessian: 1st order finite difference of autodiff Hessian 

Allows for MLE standard errors in the unconstrained space 

#### How to Verify:

```
$ ./runTests.py src/test/unit/agrad/finite_difference_test.cpp
```

#### Side Effects:

None.

#### Documentation:

None in the manual. Added doxygen documentation to agrad::finite_diff_gradient, agrad::finite_diff_hessian, agrad::finite_diff_hessian_auto, and agrad::finite_diff_grad_hessian_auto.

#### Reviewer Suggestions:

Anyone

### CODE REVIEW

**Comment by syclik** [comment]
Jenkins, retest this please.

**Comment by syclik** [comment]
@rtrangucci, want to fix the merge conflicts.

**Comment by rtrangucci** [comment]
Yessir, will do

Rob

File: src/stan/math/prim/mat/functor/finite_diff_gradient.hpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+     * <p>The functor must implement 
+     * 
+     * <code>
+     * double
+     * operator()(const
+     * Eigen::Matrix<double,Eigen::Dynamic,1>&)
+     * </code>
+     *
+     * Error should be on order of epsilon ^ 6
+     *
```
Comment (syclik) [inline src/stan/math/prim/mat/functor/finite_diff_gradient.hpp]: include a reference for this algorithm

Comment (syclik) [inline src/stan/math/prim/mat/functor/finite_diff_gradient.hpp]: and the number of function calls to calculate the finite diff gradient using this method

File: src/stan/math/prim/mat/functor/finite_diff_gradient.hpp [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+
+      int d = x.size();
+      grad_fx.resize(d);
+
+      fx = f(x);
+      
+      for (int i = 0; i < d; ++i){
+        double delta_f = 0.0;
+
+        x_temp(i) += 3.0 * epsilon;
```
Comment (syclik) [inline src/stan/math/prim/mat/functor/finite_diff_gradient.hpp]: for clarity, perhaps use:
x_temp(i) = x(i) + 3.0 * epsilon;

this will match the other calls as well.

File: src/stan/math/prim/mat/functor/finite_diff_gradient.hpp [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+        delta_f -= 45.0 * f(x_temp);
+
+        delta_f /= 60 * epsilon;
+
+        x_temp(i) = x(i);
+        grad_fx(i) = delta_f;
+      }
+
+    }
+	}
```
Comment (syclik) [inline src/stan/math/prim/mat/functor/finite_diff_gradient.hpp]: fix spacing

File: src/stan/math/prim/mat/functor/finite_diff_hessian.hpp [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+     *
+     * <p>The functor must implement 
+     * 
+     * <code>
+     * double
+     * operator()(const
+     * Eigen::Matrix<double,Eigen::Dynamic,1>&)
+     * </code>
+     *
+     * Error should be on order of epsilon ^ 4
```
Comment (syclik) [inline src/stan/math/prim/mat/functor/finite_diff_hessian.hpp]: reference?

File: src/stan/math/rev/mat/functor/finite_diff_grad_hessian.hpp [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+     * @param[in] f Function
+     * @param[in] x Argument to function
+     * @param[out] fx Function applied to argument
+     * @param[out] grad_hess_fx gradient of Hessian of function at argument
+     * @param[in] epsilon perturbation size
+     */
+    template <typename F>
+    void
+    finite_diff_grad_hessian(const F& f,
+														 const Eigen::Matrix<double,Eigen::Dynamic,1>& x,
```
Comment (syclik) [inline src/stan/math/rev/mat/functor/finite_diff_grad_hessian.hpp]: fix spacing. do you have tabs in here?

File: src/stan/math/prim/mat/functor/finite_diff_hessian.hpp [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+      int d = x.size();
+
+      Matrix<double,Dynamic,1> x_temp(x);
+      hess_fx.resize(d, d);
+
+      fx = f(x);
+      
+      double f_diff(0.0);
+      for (int i = 0; i < d; ++i){
```
Comment (syclik) [inline src/stan/math/prim/mat/functor/finite_diff_hessian.hpp]: insert space between ) and {

File: src/stan/math/prim/mat/functor/finite_diff_hessian.hpp [hunk 7]
```diff
@@ -0,0 +1,110 @@
...
+        for (int j = i; j < d; ++j){
+          x_temp(i) += 2.0 * epsilon;
+          if(i != j){
```
Comment (syclik) [inline src/stan/math/prim/mat/functor/finite_diff_hessian.hpp]: add space after if and after )

File: src/stan/math/prim/mat/functor/finite_diff_hessian.hpp [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+
+namespace stan {
+  
+  namespace math {
+
+    template <typename F>
+    double
+    finite_diff_hess_helper(const F& f,
+                            const Eigen::Matrix<double,Eigen::Dynamic,1>& x,
+                            const int& lambda,
```
Comment (syclik) [inline src/stan/math/prim/mat/functor/finite_diff_hessian.hpp]: remove &

File: src/stan/math/prim/mat/functor/finite_diff_hessian.hpp [hunk 9]
```diff
@@ -0,0 +1,110 @@
...
+     * Error should be on order of epsilon ^ 4
+     * 
+     * @tparam F Type of function
+     * @param[in] f Function
+     * @param[in] x Argument to function
+     * @param[out] fx Function applied to argument
+     * @param[out] hess_fx Hessian of function at argument
+     * @param[in] epsilon perturbation size
+     */
+
```
Comment (syclik) [inline src/stan/math/prim/mat/functor/finite_diff_hessian.hpp]: remove blank line

File: src/stan/math/rev/mat/functor/finite_diff_grad_hessian.hpp [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+      Matrix<double,Dynamic,1> x_temp(x);
+      Matrix<double,Dynamic,1> grad_auto(d);
+      Matrix<double,Dynamic,Dynamic> H_auto(d,d);
+      Matrix<double,Dynamic,Dynamic> H_diff(d,d);
+
+      
+      for (int i = 0; i < d; ++i){
+        H_diff.setZero();
+
+        x_temp(i) += 2.0 * epsilon;
```
Comment (syclik) [inline src/stan/math/rev/mat/functor/finite_diff_grad_hessian.hpp]: change to x_temp(i) = x(i) + 2.0 * epsilon;

File: src/stan/math/rev/mat/functor/finite_diff_grad_hessian.hpp [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+#include <stan/math/mix/mat/functor/gradient.hpp>
+
+namespace stan {
+  
+  namespace agrad {
+
+    /** 
+     * Calculate the value and the gradient of the hessian of the specified
+     * function at the specified argument using second-order autodiff and
+       * first-order finite difference.
```
Comment (syclik) [inline src/stan/math/rev/mat/functor/finite_diff_grad_hessian.hpp]: fix spacing

Comment (syclik) [inline src/stan/math/rev/mat/functor/finite_diff_grad_hessian.hpp]: spacing

File: src/stan/math/rev/mat/functor/finite_diff_grad_hessian.hpp [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+namespace stan {
+  
+  namespace agrad {
+
+    /** 
+     * Calculate the value and the gradient of the hessian of the specified
+     * function at the specified argument using second-order autodiff and
+       * first-order finite difference.  
+     *
+     * <p>The functor must implement
```
Comment (syclik) [inline src/stan/math/rev/mat/functor/finite_diff_grad_hessian.hpp]: include reference for the algorithm. Also indicate number of calls to f

File: src/stan/math/rev/mat/functor/finite_diff_hessian_AD.hpp [hunk 13]
```diff
@@ -0,0 +1,80 @@
...
+     * first-order finite difference.  
+     *
+     * <p>The functor must implement 
+     * 
+     * <code>
+     * double
+     * operator()(const
+     * Eigen::Matrix<double,Eigen::Dynamic,1>&)
+     * </code>
+     *
```
Comment (syclik) [inline src/stan/math/rev/mat/functor/finite_diff_hessian_AD.hpp]: include number of gradient calls

File: src/stan/math/rev/mat/functor/finite_diff_hessian_AD.hpp [hunk 14]
```diff
@@ -0,0 +1,80 @@
+#ifndef STAN__MATH__REV__MAT__FUNCTOR__FINITE_DIFF_HESSIAN_AD_HPP
```
Comment (syclik) [inline src/stan/math/rev/mat/functor/finite_diff_hessian_AD.hpp]: any reason this isn't just stan::agrad::finite_diff_hessian?

**Comment by syclik** [comment]
looks pretty good. Just fix up the things I commented on and I think we're good to go. Anyone else want to take a look?

File: src/stan/math/prim/mat/functor/finite_diff_hessian.hpp (hunk 8 above)
Comment (bob-carpenter) [inline src/stan/math/prim/mat/functor/finite_diff_hessian.hpp]: Specifically, replace "const int&" with "int".   If
it's a const ref, then it's not being used as a return,
so passing by value is the idiomatic way to go --- it should
be passing something no larger than the reference on 64-bit
architectures.

And then the const doesn't add anything to the external
API, becuase it'll be passed by value.

- Bob

File: src/stan/math/rev/mat/functor/finite_diff_grad_hessian.hpp (hunk 12 above)
Comment (bob-carpenter) [inline src/stan/math/rev/mat/functor/finite_diff_grad_hessian.hpp]: What do you mean by a reference?  As in to a book or paper?
I don't think that's necessary.  A definition of how it works
would be better.  

- Bob

**Comment by bob-carpenter** [comment]
I'd like to see tests for boundary conditions (e.g., functions with 0 and 1 arguments).

In practice, the testing of this code and autodiff will reinforce each other.

Looks OK to me.  I now see why you wanted a reference --- that's a complicated algorithm!

**Comment by rtrangucci** [comment]
@bob-carpenter Should I make a nullary functor to test the zero argument case? If so, how?

**Comment by bob-carpenter** [comment]
I don't think that makes sense.  You mean for
our positive_infinity() and negative_infinity() cases?

I do think we should be testing things like sum()
on zero-size arrays/matrices.   Some functions like
mean() and sd() will have expected failures, but sum()
should return 0.  Obviously no gradients to test!

- Bob

**Comment by rtrangucci** [comment]
@bob-carpenter Got it, I'll take care of writing these tests.

**Comment by rtrangucci** [comment]
@bob-carpenter I added argument length 0 and length 1 tests for all the finite diff functionals. They're in finite_diff_test.cpp. 

The behavior (like all of the autodiff-analogue functionals) is to return a 1-dimensional gradient, hessian and gradient of the hessian for functions with args of length 1 and a 0-dimensional gradient, hessian and gradient of the hessian (or an empty Eigen::VectorXd, MatrixXd, and vector of MatrixXd) for functions with arguments of length 0.

**Comment by bob-carpenter** [comment]
Not sure why that was directed to me, but it's the
right thing to do --- test boundary conditions, that is.

- Bob

**Comment by rtrangucci** [comment]
You asked me to add these tests about a month ago, finally got around to adding. The branch should be fine to merge now. 

Rob

**Comment by rtrangucci** [comment]
@bob-carpenter any chance you'd be able to review / merge this?

**Comment by bob-carpenter** [comment]
Sorry --- I'm used to Daniel managing all the traffic.
I merged it.

I think we also want to delete the branch at this point,
but I didn't do that.

- Bob
